### PR TITLE
Fix issue 15104

### PR DESCRIPTION
**Pull request by perillo**
The first commit fixes https://github.com/ziglang/zig/issues/15104.

The second commit was added in order to test that the bug is fixed.  Note that there is an open duplicate: #15112.

### CODE REVIEW

**Comment by chrboesch** [comment]
You get the same error as I do, must be due to the build templates.

**Comment by perillo** [comment]
Sorry, I forgot to set `is_test_command` in the tests.  I ran the test locally and it was ok but I forgot to rebuild the compiler and clean the local cache.

**Comment by chrboesch** [comment]
I'm going to close my PR now, we don't need twice.

File: lib/std/Build/RunStep.zig [hunk 1]
```diff
@@ -72,6 +72,10 @@ captured_stderr: ?*Output = null,
 
 has_side_effects: bool = false,
 
+/// Set this to true if RunStep runs a zig test.
+/// See https://github.com/ziglang/zig/issues/15104.
```
Comment (Vexu) [inline lib/std/Build/RunStep.zig:76]: I don't think this should link to an issue that the PR closes.

Comment (perillo) [reply lib/std/Build/RunStep.zig:76]: > I don't think this should link to an issue that the PR closes.

I can improve the documentation, but I'm not sure how to explain why the new that is necessary.

An alternative solution is to simply **never** close `stdin`, and instead ignore it.  However I'm not not sure if there is a valid reason for the current implementation.

Thanks.

**Review by Vexu — APPROVED** [review]

**Comment by perillo** [comment]
@Vexu Unfortunately I think this bug can also be triggered, theoretically, from a `zig init-exe` generated `build.zig` file, when the user run `zig build run`.

Here is an example (in this case the parent is `zig build`): https://github.com/ziglang/zig/issues/15104#issuecomment-491880967.

**Comment by matu3ba** [comment]
I dont understand why use cases like 
```
/// This is necessary to prevent the first child process to close stdin, and
/// the grandchild process to make stdin a pipe, causing the test executable to
/// fail to spawn with a NotDir error.
```
are not cleanly solved by using a dedicated pipe for communication, because relying on inheritable descriptors is a very bad idea and with `clone` will create reusage of otherwise taken descriptors.

If we instead use a fresh descriptor,  have 1. no problem because even it is leaked it won't be reused and 2. it scales arbitrarily for nested and multi-threaded process spawns.

PR for discussion (bitrotted a bit because nobody bothered reviewing it): #14152

**Comment by perillo** [comment]
@matu3ba I agree. The `--listen=-` option should be replaced with `--listen=fd=n,m`

By the way, since it seems that it will take some time to solve this issue, I decided to ask for #15190 to be merged.

**Comment by andrewrk** [comment]
Superseded by #15245; landed in 1728d92f60d4e9aa10d878e3235fc63764d3909b.

**Pull request closed without merging**